src/libraries/System.Net.Http/src/System/Net/Http/Metrics/MetricsHandler.cs

```diff
@@ -83,67 +80,90 @@ protected override void Dispose(bool disposing)
-                {
-                    TimeSpan duration = Stopwatch.GetElapsedTime(startTimestamp, Stopwatch.GetTimestamp());
-                    _requestsDuration.Record(duration.TotalSeconds, tags);
-                }
-
-                if (recordFailedRequests)
-                {
-                    _failedRequests.Add(1, tags);
-                }
+                _requestsDuration.Record(durationTime.TotalSeconds, tags);
             }
             else
             {
-                enrichmentContext.RecordWithEnrichment(request, response, exception, startTimestamp, tags, recordRequestDuration, recordFailedRequests, _requestsDuration, _failedRequests);
+                enrichmentContext.RecordDurationWithEnrichment(request, response, exception, durationTime, tags, _requestsDuration);
+            }
+        }
+
+        private static string GetErrorReason(Exception exception)
+        {
+            if (exception is OperationCanceledException)
+            {
+                return "cancellation";
             }
+            else if (exception is HttpRequestException e)
+            {
+                Debug.Assert(Enum.GetValues<HttpRequestError>().Length == 12, "We need to extend the mapping in case new values are added to HttpRequestError.");
+                return e.HttpRequestError switch
+                {
+                    HttpRequestError.NameResolutionError => "name_resolution_error",
+                    HttpRequestError.ConnectionError => "connection_error",
+                    HttpRequestError.SecureConnectionError => "secure_connection_error",
+                    HttpRequestError.HttpProtocolError => "http_protocol_error",
+                    HttpRequestError.ExtendedConnectNotSupported => "extended_connect_not_supported",
+                    HttpRequestError.VersionNegotiationError => "version_negotiation_error",
+                    HttpRequestError.UserAuthenticationError => "user_authentication_error",
+                    HttpRequestError.ProxyTunnelError => "proxy_tunnel_error",
+                    HttpRequestError.InvalidResponse => "invalid_response",
+                    HttpRequestError.ResponseEnded => "response_ended",
+                    HttpRequestError.ConfigurationLimitExceeded => "configuration_limit_exceeded",
+                    _ => "_OTHER"
+                };
+            }
+            return "_OTHER";
         }
 
-        private static string GetProtocolName(Version httpVersion) => (httpVersion.Major, httpVersion.Minor) switch
+        private static string GetProtocolVersionString(Version httpVersion) => (httpVersion.Major, httpVersion.Minor) switch
         {
-            (1, 0) => "HTTP/1.0",
-            (1, 1) => "HTTP/1.1",
-            (2, 0) => "HTTP/2",
-            (3, 0) => "HTTP/3",
-            _ => $"HTTP/{httpVersion.Major}.{httpVersion.Minor}"
+            (1, 0) => "1.0",
+            (1, 1) => "1.1",
+            (2, 0) => "2.0",
```

Comment:
I think @JamesNK uses "2" and "3" for HTTP/2 and HTTP/3 for asp.net core and kestrel, which, I believe, matches their RFCs ([HTTP/2](https://datatracker.ietf.org/doc/html/rfc7540), [HTTP/3](https://www.rfc-editor.org/info/rfc9114)).

can we still unify?

Reply:
Yup I will open a follow-up PR.